File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/RoomFinderDetailsActivity.java
package de.tum.in.tumcampusapp.component.tumui.roomfinder;

import android.content.ActivityNotFoundException;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.appcompat.app.AlertDialog;
import android.view.Menu;
import android.view.MenuItem;

import com.google.common.base.Optional;

import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.api.app.ApiHelper;
import de.tum.in.tumcampusapp.api.app.TUMCabeClient;
import de.tum.in.tumcampusapp.component.other.generic.ImageViewTouchFragment;
import de.tum.in.tumcampusapp.component.other.generic.activity.ActivityForLoadingInBackground;
import de.tum.in.tumcampusapp.component.other.locations.LocationManager;
import de.tum.in.tumcampusapp.component.other.locations.model.Geo;
import de.tum.in.tumcampusapp.component.tumui.roomfinder.model.RoomFinderCoordinate;
import de.tum.in.tumcampusapp.component.tumui.roomfinder.model.RoomFinderMap;
import de.tum.in.tumcampusapp.component.tumui.roomfinder.model.RoomFinderRoom;
import de.tum.in.tumcampusapp.utils.Const;
import de.tum.in.tumcampusapp.utils.NetUtils;
import de.tum.in.tumcampusapp.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Displays the map regarding the searched room.
 */
public class RoomFinderDetailsActivity
        extends ActivityForLoadingInBackground<Void, String>
        implements DialogInterface.OnClickListener {

    public static final String EXTRA_ROOM_INFO = "roomInfo";

    private ImageViewTouchFragment mImageFragment;

    private boolean mapsLoaded;

    private RoomFinderRoom room;
    private String mapId = "";
    private List<RoomFinderMap> mapsList;
    private boolean infoLoaded;

    private Fragment fragment;

    private Call<RoomFinderCoordinate> mRoomFinderCoordinateCall;
    private Call<List<RoomFinderMap>> mRoomFinderMapsCall;

    public RoomFinderDetailsActivity() {
        super(R.layout.activity_roomfinderdetails);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mImageFragment = ImageViewTouchFragment.newInstance();
        getSupportFragmentManager().beginTransaction()
                                   .add(R.id.fragment_container, mImageFragment)
                                   .commit();

        room = (RoomFinderRoom) getIntent().getSerializableExtra(EXTRA_ROOM_INFO);
        if (room == null) {
            Utils.showToast(this, "No room information passed");
            finish();
            return;
        }

        mImageFragment = ImageViewTouchFragment.newInstance();
        getSupportFragmentManager().beginTransaction()
                                   .add(R.id.fragment_container, mImageFragment)
                                   .commit();

        startLoading();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_roomfinder_detail, menu);
        MenuItem switchMap = menu.findItem(R.id.action_switch_map);
        switchMap.setVisible(!"10".equals(mapId) && mapsLoaded && fragment == null);
        MenuItem timetable = menu.findItem(R.id.action_room_timetable);
        timetable.setVisible(infoLoaded);
        timetable.setIcon(fragment == null ? R.drawable.ic_outline_event_note_24px : R.drawable.ic_outline_map_24px);
        menu.findItem(R.id.action_directions)
            .setVisible(infoLoaded && fragment == null);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int i = item.getItemId();
        if (i == R.id.action_room_timetable) {
            getRoomTimetable();
            supportInvalidateOptionsMenu();
            return true;
        } else if (i == R.id.action_directions) {
            loadGeo();
            return true;
        } else if (i == R.id.action_switch_map) {
            showMapSwitch();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onBackPressed() {
        // Remove fragment with room timetable if present and show map again
        if (fragment != null) {
            getRoomTimetable();
            supportInvalidateOptionsMenu();
            return;
        }

        super.onBackPressed();
    }

    private void getRoomTimetable() {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        // Remove if fragment is already present
        if (fragment != null) {
            ft.replace(R.id.fragment_container, mImageFragment);
            ft.commit();
            fragment = null;
            return;
        }

        String roomApiCode = room.getRoom_id();
        fragment = WeekViewFragment.newInstance(roomApiCode);
        ft.replace(R.id.fragment_container, fragment);
        ft.commit();
    }

    private void showMapSwitch() {
        CharSequence[] list = new CharSequence[mapsList.size()];
        int curPos = 0;
        for (int i = 0; i < mapsList.size(); i++) {
            list[i] = mapsList.get(i)
                              .getDescription();
            if (mapsList.get(i)
                        .getMap_id()
                        .equals(mapId)) {
                curPos = i;
            }
        }
        new AlertDialog.Builder(this).setSingleChoiceItems(list, curPos, this)
                                     .show();
    }

    @Override
    public void onClick(DialogInterface dialog, int whichButton) {
        dialog.dismiss();
        int selectedPosition = ((AlertDialog) dialog).getListView().getCheckedItemPosition();
        mapId = mapsList.get(selectedPosition).getMap_id();
        startLoading();
    }

    @Override
    protected String onLoadInBackground(Void... arg) {
        String archId = room.getArch_id();
        String url;
        if (mapId == null || mapId.isEmpty()) {
            url = Const.URL_DEFAULT_MAP_IMAGE + ApiHelper.encodeUrl(archId);
        } else {
            url = Const.URL_MAP_IMAGE + ApiHelper.encodeUrl(archId) + '/' + ApiHelper.encodeUrl(mapId);
        }
        return url;
    }

    @Override
    protected void onLoadFinished(String url) {
        mImageFragment.loadImage(url, this::showImageLoadingError);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(room.getInfo());
            getSupportActionBar().setSubtitle(room.getFormattedAddress());
        }

        showLoadingEnded();
        loadMapList();
    }

    private void loadMapList() {
        showLoadingStart();

        mRoomFinderMapsCall = TUMCabeClient.getInstance(this).fetchAvailableMaps(room.getArch_id());
        mRoomFinderMapsCall.enqueue(new Callback<List<RoomFinderMap>>() {
            @Override
            public void onResponse(@NonNull Call<List<RoomFinderMap>> call,
                                   @NonNull Response<List<RoomFinderMap>> response) {
                List<RoomFinderMap> data = response.body();
                mRoomFinderMapsCall = null;

                if (!response.isSuccessful() || data == null) {
                    onMapListLoadFailed();
                    return;
                }

                onMapListLoadFinished(Optional.of(data));
            }

            @Override
            public void onFailure(@NonNull Call<List<RoomFinderMap>> call,
                                  @NonNull Throwable throwable) {
                if (call.isCanceled()) {
                    return;
                }

                onMapListLoadFailed();
                mRoomFinderMapsCall = null;
            }
        });
    }

    private void onMapListLoadFailed() {
        onMapListLoadFinished(Optional.absent());
    }

    private void onMapListLoadFinished(Optional<List<RoomFinderMap>> result) {
        showLoadingEnded();
        if (!result.isPresent()) {
            if (NetUtils.isConnected(this)) {
                showErrorLayout();
            } else {
                showNoInternetLayout();
            }
            return;
        }
        mapsList = result.get();
        if (mapsList.size() > 1) {
            mapsLoaded = true;
        }

        supportInvalidateOptionsMenu();
    }

    private void loadGeo() {
        showLoadingStart();
        mRoomFinderCoordinateCall = TUMCabeClient.getInstance(this)
                .fetchRoomFinderCoordinates(room.getArch_id());
        mRoomFinderCoordinateCall.enqueue(new Callback<RoomFinderCoordinate>() {
            @Override
            public void onResponse(@NonNull Call<RoomFinderCoordinate> call,
                                   @NonNull Response<RoomFinderCoordinate> response) {
                RoomFinderCoordinate data = response.body();
                mRoomFinderCoordinateCall = null;

                if (!response.isSuccessful() || data == null) {
                    onLoadGeoFailed();
                    return;
                }

                onGeoLoadFinished(LocationManager.Companion.convertRoomFinderCoordinateToGeo(data));
            }

            @Override
            public void onFailure(@NonNull Call<RoomFinderCoordinate> call,
                                  @NonNull Throwable throwable) {
                if (call.isCanceled()) {
                    return;
                }

                onLoadGeoFailed();
                mRoomFinderCoordinateCall = null;
            }
        });
    }

    private void onLoadGeoFailed() {
        onGeoLoadFinished(Optional.absent());
    }

    private void onGeoLoadFinished(Optional<Geo> result) {
        showLoadingEnded();
        if (!result.isPresent()) {
            Utils.showToastOnUIThread(RoomFinderDetailsActivity.this, R.string.no_map_available);
            return;
        }

        // Build get directions intent and see if some app can handle it
        String coordinates = result.get().getLatitude() + ',' + result.get().getLongitude();
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("google.navigation:q=" + coordinates));
        List<ResolveInfo> pkgAppsList = getApplicationContext().getPackageManager()
                .queryIntentActivities(intent, PackageManager.GET_RESOLVED_FILTER);

        // If some app can handle this intent start it
        if (!pkgAppsList.isEmpty()) {
            startActivity(intent);
            return;
        }

        // If no app is capable of opening it link to google maps market entry
        try {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=com.google.android.apps.maps")));
        } catch (ActivityNotFoundException e) {
            Utils.log(e);
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/details?id=com.google.android.apps.maps")));
        }
    }

    private void showImageLoadingError() {
        if (NetUtils.isConnected(this)) {
            showError(R.string.error_something_wrong);
        } else {
            showNoInternetLayout();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();

        if (mRoomFinderMapsCall != null) {
            mRoomFinderMapsCall.cancel();
        }

        if (mRoomFinderCoordinateCall != null) {
            mRoomFinderCoordinateCall.cancel();
        }
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/RoomFinderDetailsActivity.kt
package de.tum.`in`.tumcampusapp.component.tumui.roomfinder

import android.content.ActivityNotFoundException
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.api.app.ApiHelper
import de.tum.`in`.tumcampusapp.api.app.TUMCabeClient
import de.tum.`in`.tumcampusapp.component.other.generic.ImageViewTouchFragment
import de.tum.`in`.tumcampusapp.component.other.generic.activity.ActivityForLoadingInBackground
import de.tum.`in`.tumcampusapp.component.other.locations.LocationManager
import de.tum.`in`.tumcampusapp.component.other.locations.model.Geo
import de.tum.`in`.tumcampusapp.component.tumui.roomfinder.model.RoomFinderCoordinate
import de.tum.`in`.tumcampusapp.component.tumui.roomfinder.model.RoomFinderMap
import de.tum.`in`.tumcampusapp.component.tumui.roomfinder.model.RoomFinderRoom
import de.tum.`in`.tumcampusapp.utils.Const
import de.tum.`in`.tumcampusapp.utils.NetUtils
import de.tum.`in`.tumcampusapp.utils.Utils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject

/**
 * Displays the map regarding the searched room.
 */
class RoomFinderDetailsActivity : ActivityForLoadingInBackground<Void, String>(R.layout.activity_roomfinderdetails), DialogInterface.OnClickListener {

    @Inject
    lateinit var tumCabeClient: TUMCabeClient

    private lateinit var imageFragment: ImageViewTouchFragment
    private var weekViewFragment: Fragment? = null

    private var mapsLoaded: Boolean = false

    private var mapId: String = ""
    private var mapsList: List<RoomFinderMap> = ArrayList()
    private val infoLoaded: Boolean = false

    private var roomFinderCoordinateCall: Call<RoomFinderCoordinate>? = null
    private var roomFinderMapsCall: Call<List<RoomFinderMap>>? = null

    private val room: RoomFinderRoom by lazy {
        intent.getSerializableExtra(EXTRA_ROOM_INFO) as RoomFinderRoom
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        imageFragment = ImageViewTouchFragment.newInstance()
        supportFragmentManager.beginTransaction()
                .add(R.id.fragment_container, imageFragment)
                .commit()

        startLoading()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_roomfinder_detail, menu)

        val switchMap = menu.findItem(R.id.action_switch_map)
        switchMap.isVisible = "10" != mapId && mapsLoaded && weekViewFragment == null

        val timetable = menu.findItem(R.id.action_room_timetable)
        timetable.isVisible = infoLoaded
        timetable.setIcon(
                if (weekViewFragment == null) R.drawable.ic_outline_event_note_24px
                else R.drawable.ic_outline_map_24px
        )

        menu.findItem(R.id.action_directions).isVisible = infoLoaded && weekViewFragment == null
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_room_timetable -> {
                toggleRoomTimetable()
                invalidateOptionsMenu()
                return true
            }
            R.id.action_directions -> {
                loadGeo()
                return true
            }
            R.id.action_switch_map -> {
                showMapSwitch()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onBackPressed() {
        // Remove weekViewFragment with room timetable if present and show map again
        if (weekViewFragment != null) {
            toggleRoomTimetable()
            invalidateOptionsMenu()
            return
        }

        super.onBackPressed()
    }

    private fun toggleRoomTimetable() {
        val ft = supportFragmentManager.beginTransaction()
        // Remove if weekViewFragment is already present
        if (weekViewFragment != null) {
            ft.replace(R.id.fragment_container, imageFragment)
            ft.commit()
            weekViewFragment = null
            return
        }

        val roomApiCode = room.room_id
        weekViewFragment = WeekViewFragment.newInstance(roomApiCode)
        ft.replace(R.id.fragment_container, weekViewFragment!!)
        ft.commit()
    }

    private fun showMapSwitch() {
        val descriptions = mapsList.map { it.description }
        val currentPosition = mapsList.indexOfFirst { it.map_id == mapId }

        AlertDialog.Builder(this)
                .setSingleChoiceItems(descriptions.toTypedArray(), currentPosition, this)
                .show()
    }

    override fun onClick(dialog: DialogInterface, whichButton: Int) {
        dialog.dismiss()
        val selectedPosition = (dialog as AlertDialog).listView.checkedItemPosition
        mapId = mapsList[selectedPosition].map_id
        startLoading()
    }

    override fun onLoadInBackground(vararg arg: Void): String? {
        val encodedArchId = ApiHelper.encodeUrl(room.arch_id)

        return if (mapId.isEmpty()) {
            Const.URL_DEFAULT_MAP_IMAGE + encodedArchId
        } else {
            Const.URL_MAP_IMAGE + encodedArchId + '/'.toString() + ApiHelper.encodeUrl(mapId)
        }
    }

    override fun onLoadFinished(url: String?) {
        if (url == null) {
            showImageLoadingError()
        } else {
            imageFragment.loadImage(url) { showImageLoadingError() }
        }

        supportActionBar?.title = room.info
        supportActionBar?.subtitle = room.formattedAddress

        showLoadingEnded()
        loadMapList()
    }

    private fun loadMapList() {
        showLoadingStart()

        roomFinderMapsCall = tumCabeClient.fetchAvailableMaps(room.arch_id)
        roomFinderMapsCall?.enqueue(object : Callback<List<RoomFinderMap>> {
            override fun onResponse(
                call: Call<List<RoomFinderMap>>,
                response: Response<List<RoomFinderMap>>
            ) {
                val data = response.body()
                roomFinderMapsCall = null

                if (!response.isSuccessful || data == null) {
                    onMapListLoadFailed()
                    return
                }
                onMapListLoadFinished(data)
            }

            override fun onFailure(
                call: Call<List<RoomFinderMap>>,
                throwable: Throwable
            ) {
                if (call.isCanceled) {
                    return
                }

                onMapListLoadFailed()
                roomFinderMapsCall = null
            }
        })
    }

    private fun onMapListLoadFailed() {
        onMapListLoadFinished(null)
    }

    private fun onMapListLoadFinished(result: List<RoomFinderMap>?) {
        showLoadingEnded()
        if (result == null) {
            if (NetUtils.isConnected(this)) {
                showErrorLayout()
            } else {
                showNoInternetLayout()
            }
            return
        }
        mapsList = result
        if (mapsList.size > 1) {
            mapsLoaded = true
        }

        invalidateOptionsMenu()
    }

    private fun loadGeo() {
        showLoadingStart()
        roomFinderCoordinateCall = tumCabeClient.fetchRoomFinderCoordinates(room.arch_id)
        roomFinderCoordinateCall?.enqueue(object : Callback<RoomFinderCoordinate> {
            override fun onResponse(
                call: Call<RoomFinderCoordinate>,
                response: Response<RoomFinderCoordinate>
            ) {
                val data = response.body()
                roomFinderCoordinateCall = null

                if (!response.isSuccessful || data == null) {
                    onLoadGeoFailed()
                    return
                }

                onGeoLoadFinished(LocationManager.convertRoomFinderCoordinateToGeo(data))
            }

            override fun onFailure(
                call: Call<RoomFinderCoordinate>,
                throwable: Throwable
            ) {
                if (call.isCanceled) {
                    return
                }

                onLoadGeoFailed()
                roomFinderCoordinateCall = null
            }
        })
    }

    private fun onLoadGeoFailed() {
        onGeoLoadFinished(null)
    }

    private fun onGeoLoadFinished(result: Geo?) {
        showLoadingEnded()
        if (result == null) {
            Utils.showToastOnUIThread(this@RoomFinderDetailsActivity, R.string.no_map_available)
            return
        }

        // Build get directions intent and see if some app can handle it
        val coordinates = "${result.latitude},${result.longitude}"
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("google.navigation:q=$coordinates"))
        val pkgAppsList = applicationContext.packageManager
                .queryIntentActivities(intent, PackageManager.GET_RESOLVED_FILTER)

        // If some app can handle this intent start it
        if (pkgAppsList.isNotEmpty()) {
            startActivity(intent)
            return
        }

        // If no app is capable of opening it link to google maps market entry
        try {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=com.google.android.apps.maps")))
        } catch (e: ActivityNotFoundException) {
            Utils.log(e)
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/details?id=com.google.android.apps.maps")))
        }
    }

    private fun showImageLoadingError() {
        if (NetUtils.isConnected(this)) {
            showError(R.string.error_something_wrong)
        } else {
            showNoInternetLayout()
        }
    }

    override fun onStop() {
        super.onStop()
        roomFinderMapsCall?.cancel()
        roomFinderCoordinateCall?.cancel()
    }

    companion object {
        const val EXTRA_ROOM_INFO = "roomInfo"
    }
}
